### PR TITLE
Add non-blocking opsrecipe validation

### PR DESCRIPTION
**Pull request by QuentinBisson**
Before adding a new alerting rule into this repository you should consider creating an SLO rules instead.
SLO helps you both increase the quality of your monitoring and reduce the alert noise.

* How to create a SLO rule: https://github.com/giantswarm/sloth-rules#how-to-create-a-slo
* Documentation: https://intranet.giantswarm.io/docs/monitoring/slo-alerting/

---
Towards: https://github.com/giantswarm/giantswarm/issues/23638

This PR adds a non-blocking opsrecipe validation test

### Checklist

- [x] Update CHANGELOG.md
- [x] Add [Unit tests](https://github.com/giantswarm/prometheus-rules/#testing)
- [x] Follow [Alert structure](https://github.com/giantswarm/prometheus-rules/#how-alerts-are-structured)
- [x] Consider [creating a dashboard](https://docs.giantswarm.io/getting-started/observability/grafana/custom-dashboards/) ([guidelines](https://intranet.giantswarm.io/docs/product/ux/guidelines/dashboards/)) (if it does not exist already) to help oncallers monitor the status of the issue.
- [x] Request review from oncall area, as well as team (e.g: `oncall-kaas-cloud` GitHub group).

### CODE REVIEW

**Comment by QuentinBisson** [comment]
🤔 
./test/hack/bin/check-opsrecipes.sh
git@github.com: Permission denied (publickey).
fatal: Could not read from remote repository.

File: test/hack/bin/check-opsrecipes.sh [hunk 1]
```diff
@@ -6,9 +6,11 @@ RULES_FILES=(./test/hack/output/*/*/prometheus-rules/templates/alerting-rules/*)
 #RULES_FILES=(./test/hack/output/*/prometheus-rules/templates/alerting-rules/up*)
 
 DEBUG_MODE=false
+
 CHECK_EXTRADATA_ERRORS=true
 CHECK_NORECIPE_ERRORS=true
 CHECK_UNEXISTINGRECIPE_ERRORS=true
+OPSRECIPES_DIR=./giantswarm
```
Comment (QuentinBisson) [inline test/hack/bin/check-opsrecipes.sh:13]: Ok so this is doing what I want it to but I do not like this. @hervenicol I think I need your help tomorrow

File: test/hack/bin/check-opsrecipes.sh [hunk 2]
```diff
@@ -6,9 +6,11 @@ RULES_FILES=(./test/hack/output/*/*/prometheus-rules/templates/alerting-rules/*)
 #RULES_FILES=(./test/hack/output/*/prometheus-rules/templates/alerting-rules/up*)
 
 DEBUG_MODE=false
+
```
Comment (QuentinBisson) [inline test/hack/bin/check-opsrecipes.sh:9]:
```suggestion
```

**Comment by QuentinBisson** [comment]
@giantswarm/team-atlas and specifically our bash master, what do you think about this approach?

File: Makefile.custom.mk [hunk 3]
```diff
@@ -24,16 +19,18 @@ template-chart: install-tools ## prepare the helm chart
 	test/hack/bin/architect helm template --dir helm/prometheus-rules --dry-run
 	bash ./test/hack/bin/template-chart.sh
 
+test-rules: install-tools template-chart ## run unit tests for alerting rules
+	bash test/hack/bin/verify-rules.sh "$(test_filter)"
+
+test-rules: install-tools template-chart ## run unit tests for alerting rules
+	bash test/hack/bin/verify-rules.sh "$(test_filter)"
+
```
Comment (hervenicol) [inline Makefile.custom.mk:27]: Why do we `test-rules` twice?

Comment (QuentinBisson) [reply Makefile.custom.mk:27]: Well we shoud not, that's a bad merge

File: Makefile.custom.mk [hunk 4]
```diff
@@ -6,16 +6,11 @@ clean-dry-run: ## dry run for `make clean` - print all untracked files
 clean: ## Clean the git work dir and remove all untracked files
 	# clean stage
 	git clean -xdf -- test/hack/bin test/hack/output test/hack/checkLabels
-	git checkout -- helm/prometheus-rules/Chart.yaml
-	git checkout -- helm/prometheus-rules/values.yaml
```
Comment (hervenicol) [inline Makefile.custom.mk:10]: Did you put the cleanup back somewhere else?

Comment (QuentinBisson) [reply Makefile.custom.mk:10]: The app was moved to app-build-suite so this is not needed

Comment (hervenicol) [reply Makefile.custom.mk:10]: Oh, so this is not directly related to this PR! Fine with me!

Comment (QuentinBisson) [reply Makefile.custom.mk:10]: Nope, sorry, just found it while working on this

File: test/hack/bin/check-opsrecipes.sh [hunk 5]
```diff
@@ -27,15 +28,39 @@ isInArray () {
     return 1
 }
 
+# merge_docs () merges a Hugo docs hierarchy from a source directory (first arg) into a destination directory (second arg).
+merge_docs() {
+    if [ ! -d "$1/content/docs/." ] ; then
```
Comment (hervenicol) [inline test/hack/bin/check-opsrecipes.sh:33]: These are `test` conditions as supported by `sh`.
In other places in this script we use the `bash` syntay `[[ ... ]]`.
Could we keep it consistent and use the same all around?

Comment (QuentinBisson) [reply test/hack/bin/check-opsrecipes.sh:33]:
```suggestion
    if [[ ! -d "$1/content/docs/." ]] ; then
```

File: test/hack/bin/check-opsrecipes.sh [hunk 6]
```diff
@@ -27,15 +28,39 @@ isInArray () {
     return 1
 }
 
+# merge_docs () merges a Hugo docs hierarchy from a source directory (first arg) into a destination directory (second arg).
+merge_docs() {
+    if [[ ! -d "$1/content/docs/." ]] ; then
+        echo "Source Hugo base directory not specified or invalid (must contain content/docs)!" >&2
+    fi
+    if [ ! -d "$2/content/docs/." ] ; then
```
Comment (QuentinBisson) [inline test/hack/bin/check-opsrecipes.sh:36]:
```suggestion
    if [[ ! -d "$2/content/docs/." ]] ; then
```

File: test/hack/bin/check-opsrecipes.sh [hunk 7]
```diff
@@ -27,15 +28,39 @@ isInArray () {
...
+
+    # perform merge as done by intranet build
+    merge_docs "$privateOpsrecipesHandbookParentDirectory" "$privateOpsrecipesParentDirectory"
+
     # find all ops-recipes ".md" files, and keep only the opsrecipe name (may contain a path, like "rolling-nodes/rolling-nodes")
-    find "$tmpDir"/content/docs/support-and-ops/ops-recipes -type f -name \*.md \
-        | sed -n 's_'"$tmpDir"'/content/docs/support-and-ops/ops-recipes/\(.*\).md_\1_p'
-    rm -rf "$tmpDir"
+    find "$privateOpsrecipesParentDirectory"/content/docs/support-and-ops/ops-recipes -type f -name \*.md \
+        | sed -n 's_'"$privateOpsrecipesParentDirectory"'/content/docs/support-and-ops/ops-recipes/\(.*\).md_\1_p'
```
Comment (hervenicol) [inline test/hack/bin/check-opsrecipes.sh:62]: What about the opsrecipes named `opsrecipename/_index.md`? I don't see support for them here.

Comment (QuentinBisson) [reply test/hack/bin/check-opsrecipes.sh:63]: Turns out it's not needed anymore, this was used for 1 ops-recipe which was changed by the owning team

File: test/hack/bin/check-opsrecipes.sh [hunk 8]
```diff
@@ -57,7 +90,7 @@ main() {
     ########################
 
     # Retrieve list of opsrecipes
-    mapfile -t opsRecipes < <(listOpsRecipes)
+    mapfile -t opsRecipes < <(listOpsRecipes $runInCi)
```
Comment (hervenicol) [inline test/hack/bin/check-opsrecipes.sh:94]: :heart: quotes :heart: :grin: 
```suggestion
    mapfile -t opsRecipes < <(listOpsRecipes "$runInCi")
```

**Comment by hervenicol** [comment]
> @giantswarm/team-atlas and specifically our bash master, what do you think about this approach?

Is there something specific you want to discuss?
This PR does not change much in the logic, I don't see much to say about it.

**Comment by QuentinBisson** [comment]
Are you fine with the ci and non ci make targets? Also, it changed a bit after some thinking over the weekend :D

**Comment by QuentinBisson** [comment]
But overall, are you fine with the current script as I will add it in sloth rules as well

**Comment by hervenicol** [comment]
Well, I first thought there should be a way to auto-detect CI, but having it manual like this really outlines what is different in the CI.
I'm fine with it.

**Review by hervenicol — APPROVED** [review]
LGTN

**Pull request merged**